packages/async-flow/src/async-flow.js

```diff
@@ -0,0 +1,417 @@
+              eph.membrane.stop();
+            }
+            tmp.resetFor(flow);
+            log.reset();
+            bijection.reset();
+
+            // This is not an expected throw, so in theory arbitrary chaos
+            // may ensue from throwing it. But at this point
+            // we should have successfully isolated this activation from
+            // having any observable effects on the host, aside from
+            // console logging and
+            // resource exhaustion, including infinite loops
+            const err = makeError(
+              X`In a replay failure: see getFailures() for more information`,
+            );
+            annotateError(err, X`due to ${fatalProblem}`);
+            throw err;
+          },
+        },
+        wakeWatcher: {
+          onFulfilled(_fulfillment) {
+            const { facets } = this;
+            facets.flow.wake();
+          },
+          onRejected(_fulfillment) {
+            const { facets } = this;
+            facets.flow.wake();
+          },
+        },
+      },
+    );
+    const makeAsyncFlowKit = (activationThis, activationArgs) => {
+      const asyncFlowKit = internalMakeAsyncFlowKit(
+        activationThis,
+        activationArgs,
+      );
+      asyncFlowKit.flow.restart();
+      return asyncFlowKit;
+    };
+    return harden(makeAsyncFlowKit);
+  };
+
+  /**
+   * @param {Zone} zone
+   * @param {string} tag
+   * @param {GuestAsyncFunc} guestFunc
+   * @param {{ startEager?: boolean }} [options]
+   * @returns {HostAsyncFuncWrapper}
+   */
+  const asyncFlow = (zone, tag, guestFunc, options = undefined) => {
+    const makeAsyncFlowKit = prepareAsyncFlowKit(zone, tag, guestFunc, options);
+    const hostFuncName = `${guestFunc.name || 'anon'}_hostWrapper`;
+    const wrapperFunc = {
+      [hostFuncName](...args) {
+        const { flow } = makeAsyncFlowKit(this, args);
+        const outcomeVow = flow.getOutcome();
+        flowForOutcomeVowKey.init(vowishKey(outcomeVow), flow);
+        return outcomeVow;
+      },
+    }[hostFuncName];
+    defineProperties(wrapperFunc, {
+      length: { value: guestFunc.length },
+    });
+    return harden(wrapperFunc);
+  };
+
+  const adminAsyncFlow = outerZone.exo('AdminAsyncFlow', AdminAsyncFlowI, {
+    getFailures() {
+      return failures.snapshot();
+    },
+    wakeAll() {
+      // [...stuff.keys()] in order to snapshot before iterating
+      const failuresToRestart = [...failures.keys()];
+      const flowsToWake = [...eagerWakers.keys()];
+      for (const flow of failuresToRestart) {
+        flow.restart();
+      }
+      for (const flow of flowsToWake) {
+        flow.wake();
+      }
+    },
+    getFlowForOutcomeVow(outcomeVow) {
+      return flowForOutcomeVowKey.get(vowishKey(outcomeVow));
+    },
+  });
+
+  // Cannot call this until everything is prepared
+  // adminAsyncFlow.wakeAll();
+
+  return harden({
+    prepareAsyncFlowKit,
+    asyncFlow,
+    adminAsyncFlow,
+  });
```

Comment:
Should this return the prepared `vowTools`, so it can be used as the "vow entrypoint"?

Reply:
Let's revisit this once we're addressing https://github.com/Agoric/agoric-sdk/issues/9231 in later PRs.

Resolving